iroha/src/crypto.rs

```diff
@@ -106,6 +107,66 @@ impl Debug for Signature {
     }
 }
 
+/// Describes an object that can be signed.
+pub trait Sign {
+    /// The bytes that are signed.
+    fn payload(&self) -> Vec<u8>;
+
+    /// Method to get the signature based on the `payload`.
+    fn signature(
+        &self,
+        public_key: &PublicKey,
+        private_key: &PrivateKey,
+    ) -> Result<Signature, String> {
+        Ok(Signature::new(*public_key, &self.payload(), private_key)?)
+    }
+}
+
+/// Container for multiple signatures.
+#[derive(Debug, Clone, Encode, Decode, Default)]
+pub struct Signatures {
+    signatures: BTreeMap<PublicKey, Signature>,
+}
+
+impl Signatures {
+    /// Adds multiple signatures and skips the duplicates.
+    pub fn add_multiple(&mut self, signatures: &[Signature]) {
+        for signature in signatures.iter().cloned() {
+            let _result = self.try_add(signature);
+        }
+    }
+
+    /// Tries to add a signature. If the signature already exists - returns an error.
+    #[allow(clippy::map_entry)]
+    pub fn try_add(&mut self, signature: Signature) -> Result<(), String> {
+        if self.signatures.contains_key(&signature.public_key) {
+            Err("Already signed by these keys.".to_string())
+        } else {
+            self.signatures.insert(signature.public_key, signature);
+            Ok(())
+        }
+    }
+
+    /// Returns signatures that have passed verification.
+    pub fn verified(&self, payload: &[u8]) -> Vec<Signature> {
```

Comment:
Think about naming. Here you not just filter signatures, but have a business logic, based on the payload you get signatures that verify this payload.
Do we need to store not verified signatures at all?

Reply:
Okay, I will try to rewrite it so that it stores only verified signatures.

Reply:
Actually I just realized that verifying signatures on addition does not make sense, as the place where we want to verify signatures is when the peer receives a message and wants to check correspondence of this specific signatures for this message. But I will try to refactor it a bit anyway.